MINOR: Add System test for standby task-rebalancing

bbejeck opened the pull request:
System test for standby replicas and rebalancing occurs within a reasonable period (20 seconds) the timeout period for verification was increased due to the nature of nature of checking the rebalance.


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
MINOR: add test for standby task-rebalancing

MINOR: needed to add start up message to driver

MINOR: debugging why won't start

MINOR: increase timeouts and set correct counts

MINOR: don't need counts after restart

MINOR: don't need counts after restart, missed one

MINOR: assert consumed from both topics

MINOR: try 20 second timeouts

bbejeck commented:
@guozhangwang @mjsax @dguy

bbejeck commented:
[System test results](http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-02-09--001.1518151479--bbejeck--MINOR_system_tests_for_standby_tasks--30f8e0a/report.html)

dguy commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,224 @@
+        builder.addStateStore(storeBuilder);
+        builder.addStateStore(persistentStoreBuilder);
+
+        streamOne.transformValues(new StandbyTransformValuesSupplier(inMemoryStoreName), inMemoryStoreName)
```
any reason why you don't use `builder.table(..)` instead of `transformValues(..)` ?

bbejeck replied:
not sure, but will switch

EDIT: I remember now, I wanted to originally modify the values, but not needed anymore.

dguy commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,159 @@
+        self.zk.start()
+
+    def test_standby_tasks_rebalance(self):
+        self.kafka.start()
```
`setUp` ?

bbejeck replied:
ack

bbejeck committed:
MINOR: updates per comments

bbejeck committed:
MINOR: updates per comments

bbejeck commented:
updated this

bbejeck commented:
re-running system tests

bbejeck committed:
MINOR: putting kafka start back in test

bbejeck committed:
MINOR: needed to add new input topic an increase exepected task count

bbejeck committed:
MINOR: changing to use groupByKey to re-use same stream with different store types

bbejeck committed:
MINOR: checkstyle error fix

bbejeck commented:
updated [system test results](http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-02-12--001.1518450607--bbejeck--MINOR_system_tests_for_standby_tasks--1b21fcd/report.html)

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+        // it is expected that max.poll.interval, retries, request.timeout and max.block.ms set
+        // streams_broker_down_resilience_test and passed as args
+        if (additionalConfigs != null && !additionalConfigs.equalsIgnoreCase("none")) {
+            Map<String, String> updated = updatedConfigs(additionalConfigs);
```
nit: add `final`

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        if (!confirmCorrectConfigs(streamsProperties)) {
+            System.err.println(String.format("ERROR: Did not have all required configs expected  to contain %s %s %s %s",
```
Nit: double space -- should we add `,` in between `%s, %s, ...` ?

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+
+        ValueMapper<Long, String> countMapper = new ValueMapper<Long, String>() {
+            @Override
+            public String apply(Long value) {
```
nit: `final`

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+
+        streams.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+            @Override
+            public void uncaughtException(final Thread t, final Throwable e) {
```
nit: `final`

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+
+        streams.setStateListener(new KafkaStreams.StateListener() {
+            @Override
+            public void onChange(KafkaStreams.State newState, KafkaStreams.State oldState) {
```
nit: `final` (also 2x in the body)

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+        streams.close(10, TimeUnit.SECONDS);
+    }
+
+    private static boolean confirmCorrectConfigs(Properties properties) {
```
nit: `final`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+     * @param formattedConfigs the formatted config string
+     * @return HashMap with keys and values inserted
+     */
+    private static Map<String, String> updatedConfigs(String formattedConfigs) {
```
nit: `final` (couple more in the body)

bbejeck replied:
ack

guozhangwang commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,159 @@
+                                      self.output_topic_2: {'partitions': 1, 'replication-factor': 1}
+                                  })
+
+    def get_consumer(self, topic, num_messages):
```
Could we reuse the function `get_consumer` in `StreamsBrokerDownResilience`?

bbejeck replied:
Ack, for this comment and below, I was considering adding a Jira to fold several parts of the helper methods into `streams.py` for re-use across other tests.

WDYT?
\cc @dguy @mjsax

guozhangwang replied:
Sounds good.

mjsax replied:
👍

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandbyReplicaDriver.java:
```diff
@@ -0,0 +1,77 @@
+
+import java.util.Properties;
+
+public class StreamsStandbyReplicaDriver {
```
I think we can rename this class and the `StreamsStandbyTaskDriverService` below more generally since it could potentially be shared in other tests in the future. Since what it does is just keep sending messages to the broker until enough has been sent or being told to stop. For example, we can just create in `tests/kafkatest/services/simple_producer.py` for this producer?

bbejeck replied:
ack, see comment above.

guozhangwang commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,159 @@
+
+    def test_standby_tasks_rebalance(self):
+
+        driver = StreamsStandbyTaskDriverService(self.test_context, self.kafka, self.num_messages)
```
Should we verify this driver is closed (i.e. specified num_messages have been sent) in the end before verification?

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandbyReplicaDriver.java:
```diff
@@ -0,0 +1,77 @@
+        producerProps.put(ProducerConfig.CLIENT_ID_CONFIG, "StandbyTaskTestsProducer");
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
```
We should enable idempotency?

Also compare `EosTestDriver` on how to handle `TimeoutException` -- as we assume a fixed number of messages to be produced, we must insure we don't loose data when producing.

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandbyReplicaDriver.java:
```diff
@@ -0,0 +1,77 @@
+        try (final KafkaProducer<String, String> kafkaProducer = new KafkaProducer<>(producerProps)) {
+
+            while (keepProducing && messageCounter++ < numMessages) {
+                for (String topic : topics) {
```
nit: `final`

bbejeck replied:
ack

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+    static final String SOURCE_TOPIC = "standbyTaskSource1";
+    static final String SOURCE_TOPIC_2 = "standbyTaskSource2";
+
+    private static final String SINK_TOPIC_1 = "standbyTaskSink1";
```
These constant string names are scattered in the python code and java code. Could we let python code pass in them as parameters and let this class as well as the `StreamsStandbyReplicaDriver` code read them accordingly to make such constants concentrated in one place?

bbejeck replied:
ack

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,184 @@
+
+
+    static final String SOURCE_TOPIC = "standbyTaskSource1";
+    static final String SOURCE_TOPIC_2 = "standbyTaskSource2";
```
`standbyTaskSource2` is not specified in the python code and hence not created?

bbejeck replied:
ack, need to remove it.

guozhangwang reviewed (commented):
Left some comments.

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,159 @@
+
+    @staticmethod
+    def verify_from_file(processor, message, file):
+        result = processor.node.account.ssh_output("grep '%s' %s | wc -l" % (message, file), allow_fail=False)
```
Why do you not use the `account.monitor_log` but implement your own method?

bbejeck replied:
to avoid timing issues.  `account.monitor_log` is best used immediately after some action, but in this test, there are several instances where I want to go back and verify behavior by looking at the entire log file.  I've also encountered failures with `monitor_log` and when going back to the `stdout` file and see the expected result in the file.  

 Note I think this approach could help EOS tests.

mjsax replied:
Thanks for pointing out. Wondering if we should improve `monitor_log` directly?

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,159 @@
+
+class StreamsStandbyTask(Test):
+    """
+    This test validates using standby tasks helps with rebalance times
```
> helps with rebalance times

How is this validated?

bbejeck replied:
the timeout for verifying the transition from `REBALANCING to RUNNING` is set somewhat strict (20 seconds)

bbejeck committed:
updates and refactoring per comments

guozhangwang commented:
@bbejeck Did you miss additional commits addressing the comments? From the comments you seems have already addressed them, so just checking if you just forgot to `git push` :P

bbejeck commented:
@guozhangwang haven't finished yet, I'll push later on.

bbejeck committed:
MINOR: updates per comments, remove redundant tests, unused methods

bbejeck committed:
MINOR: checkstyle fixes

bbejeck commented:
rebased and updated per comments. Kicking off new system test, [results here](http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-02-17--001.1518826500--bbejeck--MINOR_system_tests_for_standby_tasks--46eb570/report.html)

Note that the comments for refactoring for common usage will be addressed in a follow-up PR.

bbejeck committed:
MINOR: formatting fixes

bbejeck committed:
MINOR: need to take out brackets

guozhangwang commented:
LGTM. Ping @mjsax for another look before merging.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsRepeatingIntegerKeyProducer.java:
```diff
@@ -0,0 +1,108 @@
+                                    try {
+                                        // message == org.apache.kafka.common.errors.TimeoutException: Expiring 4 record(s) for data-0: 30004 ms has passed since last attempt plus backoff time
+                                        final int expired = Integer.parseInt(exception.getMessage().split(" ")[2]);
+                                        messageCounter -= expired;
```
Should messageCounter be `volatile`? Not sure if this callback happens in the same or a different thread? \cc @guozhangwang

bbejeck replied:
Ack, adding as I think callback happens on different thread

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsRepeatingIntegerKeyProducer.java:
```diff
@@ -0,0 +1,108 @@
+                                        // message == org.apache.kafka.common.errors.TimeoutException: Expiring 4 record(s) for data-0: 30004 ms has passed since last attempt plus backoff time
+                                        final int expired = Integer.parseInt(exception.getMessage().split(" ")[2]);
+                                        messageCounter -= expired;
+                                    } catch (Exception ignore) {
```
nit: add `final`

bbejeck replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/StreamsStandByReplicaTest.java:
```diff
@@ -0,0 +1,161 @@
+        streamsProperties.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100);
+        streamsProperties.put(StreamsConfig.STATE_DIR_CONFIG, stateDirStr);
+        streamsProperties.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 1);
+        streamsProperties.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
```
Should we increase retries and enable idempotent producer?

bbejeck replied:
ack. Retries are set via configs sent to test, currently set to 2 increase more?

mjsax replied:
Ack. Should be ok than.

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+
+        processor_1.start()
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:3 STANDBY_TASKS:3", processor_1.STDOUT_FILE)
```
should this be in the file twice? cf L115? (note the latest change that does not delete log files anymore if a processor is stopped -- or enable cleanup for this test)

mjsax reviewed (commented):
Some more comments and questions.

bbejeck committed:
MINOR: updates per comments

bbejeck committed:
MINOR: adding counts since clean is not enabled

bbejeck commented:
updated this
did a rebase to pick up the clean option on shutdown and updated the test to reflect no clean-up is specified.

Kicked off [new system test](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1330/)

bbejeck committed:
MINOR: putting original counts back

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+
+        processor_2.start()
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_1.STDOUT_FILE)
```
Should this be `num_lines=2` (cf. L120)

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+        processor_2.start()
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_1.STDOUT_FILE)
+        self.wait_for_verification(processor_2, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_2.STDOUT_FILE)
```
Should this be `num_lines=2` (cf. L121)

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+
+        processor_3.stop()
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:3 STANDBY_TASKS:3", processor_1.STDOUT_FILE, num_lines=2)
```
Should this be `num_lines=3` (cf. L115 and L135)

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+        processor_3.stop()
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:3 STANDBY_TASKS:3", processor_1.STDOUT_FILE, num_lines=2)
+        self.wait_for_verification(processor_2, "ACTIVE_TASKS:3 STANDBY_TASKS:3", processor_2.STDOUT_FILE)
```
Should this be `num_lines=3` (cf. L116 and L126)

bbejeck committed:
MINOR: missed one changed count

mjsax commented on tests/kafkatest/tests/streams/streams_standby_replica_test.py:
```diff
@@ -0,0 +1,166 @@
+
+        self.wait_for_verification(processor_1, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_1.STDOUT_FILE)
+        self.wait_for_verification(processor_3, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_3.STDOUT_FILE)
+        self.wait_for_verification(processor_2, "ACTIVE_TASKS:2 STANDBY_TASKS:2", processor_2.STDOUT_FILE, num_lines=2)
```
Should this be `num_lines=3` for all three (cf. L140-L142)

bbejeck replied:
For this comment and others for the expected line count, I've posted latest results below.

mjsax reviewed (commented):
Some more comments

bbejeck commented:
as per the line counts these two builds failed 
 https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1330/
  https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1331/
Trying with original counts and see what happens

bbejeck commented:
Latest build https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1332/ with original counts passes

mjsax reviewed (approved):
Ah. Just double checked the code. Your test inherits from `StreamsTestBaseService` but we disabled intermediate cleanup only in `StreamsEosTestBaseService`...

LGTM.

bbejeck commented:
ping @mjsax @guozhangwang for final review and merging
